packages/uicore/src/components/StringSubstitute/StringSubstitute.tsx

```diff
@@ -1,49 +1,63 @@
 import React, { Fragment } from 'react'
 
 type SubstituteVars = Record<string, any> // eslint-disable-line @typescript-eslint/no-explicit-any
+const MATCH_ELSE_KEY = '___'
+const MAX_LOOP_COUNT = 1000
 
 function translateExpression(str: string, key: string, vars: SubstituteVars) {
+  let startFrom = 0
+  let loopCount = 0 // to prevent complex strings that may cause infinite loop
+
   // Replace simple i18n expression {key|match1:value1,match2:value2}
   // Sample: '{user} wants to merge {number} {number|1:commit,commits} into {target} from {source}'
   // Find `{number|`
-  const startIndex = str.indexOf(`{${key}|`)
-  const MATCH_ELSE_KEY = '___'
+  do {
+    loopCount++
+    const startIndex = str.indexOf(`{${key}|`, startFrom)
+
+    if (startIndex === -1) {
+      return str
+    }
 
-  if (startIndex !== -1) {
     // Find closing `}`
     const endIndex = str.indexOf('}', startIndex)
 
-    if (endIndex !== -1) {
-      // Get whole expression of `{number|1:commit,commits}`
-      const expression = str.substring(startIndex, endIndex + 1)
+    if (endIndex === -1) {
+      return str
+    }
+
+    // Get whole expression of `{number|1:commit,commits}`
+    const expression = str.substring(startIndex, endIndex + 1)
 
-      // Build value mapping from expression
-      const mapping = expression
-        .split('|')[1] // Get `1:commit,commits}`
-        .slice(0, -1) // Remove last closing `}`
-        .split(',') // ['1:commit', 'commits']
-        .reduce((map, token) => {
-          // Convert to a map { 1: commit, [MATCH_ELSE_KEY]: commits }
-          const [k, v] = token.split(':')
-          map[v ? k : MATCH_ELSE_KEY] = v || k
-          return map
-        }, {} as Record<string, string>)
+    // Build value mapping from expression
+    const mapping = expression
+      .split('|')[1] // Get `1:commit,commits}`
+      .slice(0, -1) // Remove last closing `}`
+      .split(',') // ['1:commit', 'commits']
+      .reduce((map, token) => {
+        // Convert to a map { 1: commit, [MATCH_ELSE_KEY]: commits }
+        const [k, v] = token.split(':')
+        map[v ? k : MATCH_ELSE_KEY] = v || k
+        return map
+      }, {} as Record<string, string>)
 
-      const matchedValue = mapping[vars[key]] || mapping[MATCH_ELSE_KEY]
+    const matchedValue = mapping[vars[key]] || mapping[MATCH_ELSE_KEY]
 
-      if (matchedValue) {
-        return str.replace(expression, matchedValue)
-      }
+    if (matchedValue) {
+      startFrom = startIndex + matchedValue.length
+      str = str.replace(expression, matchedValue)
+    } else {
+      startFrom = endIndex
     }
-  }
+  } while (startFrom < str.length || loopCount > MAX_LOOP_COUNT)
```

Comment:
I'm pretty sure that should be `startFrom < str.length && loopCount <= MAX_LOOP_COUNT` (we want to stop once we reached the MAX. *Note* the `=` is required since the counter is incremented at the beginning of the loop, so we start with 1 and thus the MAX is inclusive)

Reply:
Good eyes. Fixed